### ISSUE TITLE
pl.Expr.any() produces wrong results in windows functions

### ISSUE THREAD

**Issue by cbilot** [pola-rs/polars#2676]
Linux Mint 20.3
python 3.10.2
polars 0.13.3

#### Describe your bug.
When the pl.Expr.any() expression is used over a window function, it produces incorrect results.

**Caution**: the results are **non-deterministic**.  Using the same starting data, the pl.Expr.any() expression produces different answers with each run (and this is not a result of using a random number generator to create the initial dataset).
#### What are the steps to reproduce the behavior?
```python
import random

import polars as pl

# Generate an initial dataset with two rows per row_id and a random selection of True/False values for each row
df = (
    pl.DataFrame(
        {
            "row_id": list(range(0, 500)) * 2,
            "boolvar": [random.getrandbits(1) for x in range(0, 1000)],
        }
    )
    .sort(["row_id"])
    .with_column(pl.col("boolvar").cast(pl.datatypes.Boolean).keep_name())
)
print(df)
```
```
┌────────┬─────────┐
│ row_id ┆ boolvar │
│ ---    ┆ ---     │
│ i64    ┆ bool    │
╞════════╪═════════╡
│ 0      ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 0      ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 1      ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 1      ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...    ┆ ...     │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 498    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 498    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 499    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 499    ┆ false   │
└────────┴─────────┘
```

Now, run a windowed pl.Expr.any() expression -- twice -- on the **same** data...
```python
pl.testing.assert_frame_equal(
    df.with_column(pl.any("boolvar").over("row_id").alias("result")),
    df.with_column(pl.any("boolvar").over("row_id").alias("result")),
)
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/corey/.virtualenvs/test_project3.10/lib/python3.10/site-packages/polars/testing.py", line 99, in assert_frame_equal
    _assert_series_inner(
  File "/home/corey/.virtualenvs/test_project3.10/lib/python3.10/site-packages/polars/testing.py", line 183, in _assert_series_inner
    raise_assert_detail(
  File "/home/corey/.virtualenvs/test_project3.10/lib/python3.10/site-packages/polars/testing.py", line 209, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame are different
```
#### What is the actual behavior?
Let's take a closer look at any pairs of row_id's for which the windowed pl.Expr.any() expression produced different values:
```python
df = df.with_column(pl.any("boolvar").over("row_id").alias("result1"))
df = df.with_column(pl.any("boolvar").over("row_id").alias("result2"))

diffs = df.filter((pl.col("result1") != pl.col("result2")))

df.join(diffs.select(pl.col("row_id").unique()), on="row_id")
```

One my *first* run using the generated data, I receive this result:
```
shape: (62, 4)
┌────────┬─────────┬─────────┬─────────┐
│ row_id ┆ boolvar ┆ result1 ┆ result2 │
│ ---    ┆ ---     ┆ ---     ┆ ---     │
│ i64    ┆ bool    ┆ bool    ┆ bool    │
╞════════╪═════════╪═════════╪═════════╡
│ 72     ┆ false   ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 72     ┆ true    ┆ false   ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 104    ┆ false   ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 104    ┆ true    ┆ false   ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...    ┆ ...     ┆ ...     ┆ ...     │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 482    ┆ true    ┆ false   ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 482    ┆ false   ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 491    ┆ true    ┆ true    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 491    ┆ false   ┆ true    ┆ true    │
└────────┴─────────┴─────────┴─────────┘
```
Note how for some row_id's, the windowed pl.Expr.any() expression does not produce a consistent value for both records with the same row_id.

Let's run just the windowed pl.Expr.any() expressions again on the **same** data (that is, **without** regenerating any new randomized boolean values).
```python
df.drop_in_place('result1')
df.drop_in_place('result2')

df = df.with_column(pl.any("boolvar").over("row_id").alias("result1"))
df = df.with_column(pl.any("boolvar").over("row_id").alias("result2"))

diffs = df.filter((pl.col("result1") != pl.col("result2")))

df.join(diffs.select(pl.col("row_id").unique()), on="row_id")
```
Now I receive different results:
```
shape: (22, 4)
┌────────┬─────────┬─────────┬─────────┐
│ row_id ┆ boolvar ┆ result1 ┆ result2 │
│ ---    ┆ ---     ┆ ---     ┆ ---     │
│ i64    ┆ bool    ┆ bool    ┆ bool    │
╞════════╪═════════╪═════════╪═════════╡
│ 67     ┆ false   ┆ true    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 67     ┆ true    ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 131    ┆ false   ┆ true    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 131    ┆ true    ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...    ┆ ...     ┆ ...     ┆ ...     │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 450    ┆ true    ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 450    ┆ false   ┆ true    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 482    ┆ true    ┆ false   ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 482    ┆ false   ┆ true    ┆ false   │
└────────┴─────────┴─────────┴─────────┘
```
Note how: 
* the number of records involved has changed
* row_id 482 appears again, but some of the affected row_id's are different

 
#### Some other notes
* This was a **ridiculously** hard bug to recreate using test data.  I could not produce an example of this problem without resorting to generating random boolean values and generating sufficiently large datasets to increase the likelihood of getting an error.
* From my testing, the analogous pl.Expr.all() expression does not suffer from this bug.  Only the any() expression.
* I found this bug while working with boolean values that I generated using Expressions.  That is, I did not cast any values from integer to boolean.
* The error occurs even if the row_id is cast to a different type: for example, i32.  
* Complete stab in the dark: is there a missing sort of some kind when the result of the windowed any() expression is merged back into the dataset?  The fact that the windowed value is not consistent on pairs of row_id's suggests something like this.

**Comment by ritchie46** [comment]
Thanks for the issue report. Working on it!

**Comment by ritchie46** [comment]
I have got a fix in #2680. Also found that `pl.any/pl.all` is a lot slower than it has to be. That is now also improved by dispatching to the proper implementations.
